### PR TITLE
Make the notification slider work

### PR DESCRIPTION
**Pull request by dbkr**
Changes notification toggle from binary everything/mentions-only to the 4 level slider. Also uses the push rule account data to update the state rather than an internal dispatch, so the state remains consistent if it's changed from a different client.

Also fixes https://github.com/vector-im/vector-web/issues/1495

### CODE REVIEW

File: src/components/views/rooms/RoomTile.js [hunk 1]
```diff
@@ -179,15 +184,18 @@ module.exports = React.createClass({
         var notificationCount = this.props.room.getUnreadNotificationCount();
         // var highlightCount = this.props.room.getUnreadNotificationCount("highlight");
 
+        var badges = notificationCount > 0 && this._shouldShowNotifBadge();
+        badges |= this.props.highlight && this._shouldShowMentionBadge();
```
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: not a fan of the use of bitwise `|` here

File: src/RoomNotifs.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        return 'mute';
+    }
+
+    // for everything else, look at the room rule.
+    const roomRule = MatrixClientPeg.get().getRoomPushRule('global', roomId);
+
+    // XXX: We have to assume the default is to notify for all messages
+    // (in particular this will be 'wrong' for one to one rooms because
+    // they will notify loudly for all messages)
+    if (!roomRule || !roomRule.enabled) return 'all_messages';
```
Comment (richvdh) [inline src/RoomNotifs.js]: could we use some sort of enum for the state constants?

File: src/components/views/rooms/RoomTile.js [hunk 3]
```diff
@@ -43,43 +44,47 @@ module.exports = React.createClass({
...
-                }
-                break;
+    _getNotifState: function() {
+        if (MatrixClientPeg.get().isGuest()) return 'all_messages';
+        return RoomNotifs.getRoomNotifsState(this.props.room.roomId);
+    },
+
+    _shouldShowNotifBadge: function() {
+        const showBadgeInStates = ['all_messages', 'all_messages_loud'];
+        const currentState = this._getNotifState();
```
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: why are we calling `_getNotifState` again rather than using `this.state.notifState`?

File: src/components/views/rooms/RoomTile.js [hunk 4]
```diff
@@ -43,43 +44,47 @@ module.exports = React.createClass({
...
+        return showBadgeInStates.indexOf(currentState) > -1;
+    },
+
+    _shouldShowMentionBadge: function() {
+        return this._getNotifState() != 'mute';
```
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: ditto

File: src/components/views/rooms/RoomTile.js [hunk 5]
```diff
@@ -43,43 +44,47 @@ module.exports = React.createClass({
...
-    onAction: function(payload) {
-        switch (payload.action) {
-            case 'notification_change':
-                // Is the notification about this room?
-                if (payload.roomId === this.props.room.roomId) {
-                    this.setState( { areNotifsMuted : payload.areNotifsMuted });
-                }
-                break;
+    _getNotifState: function() {
+        if (MatrixClientPeg.get().isGuest()) return 'all_messages';
```
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: i'd be inclined to push this down into `getRoomNotifsState`

File: src/components/views/rooms/RoomTile.js [hunk 6]
```diff
@@ -179,15 +184,18 @@ module.exports = React.createClass({
         var notificationCount = this.props.room.getUnreadNotificationCount();
```
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: why are some notifications considered by the parent (via this.props.highlight) and some right here? It's rather confusing.

File: src/components/views/rooms/RoomTile.js [hunk 7]
```diff
@@ -179,15 +184,18 @@ module.exports = React.createClass({
...
+        badges |= this.props.highlight && this._shouldShowMentionBadge();
+
         var classes = classNames({
             'mx_RoomTile': true,
             'mx_RoomTile_selected': this.props.selected,
             'mx_RoomTile_unread': this.props.unread,
-            'mx_RoomTile_unreadNotify': notificationCount > 0 && !this.state.areNotifsMuted,
-            'mx_RoomTile_highlight': this.props.highlight,
+            'mx_RoomTile_unreadNotify': notificationCount > 0 && this._shouldShowNotifBadge(),
+            'mx_RoomTile_highlight': this.props.highlight && badges,
```
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: why `&& badges`? should this be `&& this._shouldShowMentionBadge()`?

File: src/components/views/rooms/RoomTile.js (hunk 1 above)
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: How about factoring out the sub-clauses to local vars?

File: src/RoomNotifs.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+import MatrixClientPeg from './MatrixClientPeg';
+import PushProcessor from 'matrix-js-sdk/lib/pushprocessor';
+import q from 'q';
+
+export function getRoomNotifsState(roomId) {
+    // look through the override rules for a rule affecting this room:
+    // if one exists, it will take precedence.
+    const muteRule = findOverrideMuteRule(roomId);
+    if (muteRule && muteRule.enabled) {
```
Comment (richvdh) [inline src/RoomNotifs.js]: should the check of `enabled` be done in `findOverrideMuteRule`? what if there are multiple rules which look like mute rules, of which the first is disabled and the second enabled?

File: src/RoomNotifs.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    if (actionsObject.tweaks.sound) return 'all_messages_loud';
+
+    return null;
+}
+
+export function setRoomNotifsState(roomId, newState) {
+    const cli = MatrixClientPeg.get();
+    const promises = [];
+
+    if (newState == 'mute') {
```
Comment (richvdh) [inline src/RoomNotifs.js]: split this into two functions?

File: src/RoomNotifs.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        // delete the room rule
+        const roomRule = MatrixClientPeg.get().getRoomPushRule('global', roomId);
```
Comment (richvdh) [inline src/RoomNotifs.js]: s/MatrixClientPeg.get()/cli/

File: src/RoomNotifs.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        if (roomRule) {
+            promises.push(cli.deletePushRule('global', 'room', roomRule.rule_id));
+        }
+
+        // add an override rule to squelch everything in this room
+        promises.push(cli.addPushRule('global', 'override', roomId, {
```
Comment (richvdh) [inline src/RoomNotifs.js]: will this behave correctly if there is already a room rule?

File: src/components/views/rooms/RoomTile.js (hunk 6 above)
Comment (dbkr) [inline src/components/views/rooms/RoomTile.js]: Good question :/

File: src/RoomNotifs.js (hunk 11 above)
Comment (dbkr) [inline src/RoomNotifs.js]: Yep, there can only ever be one room rule so it will just be replaced.

File: src/RoomNotifs.js [hunk 12]
```diff
@@ -18,12 +18,19 @@ import MatrixClientPeg from './MatrixClientPeg';
 import PushProcessor from 'matrix-js-sdk/lib/pushprocessor';
 import q from 'q';
 
+export const ALL_MESSAGES_LOUD = 'all_messages_loud';
```
Comment (richvdh) [inline src/RoomNotifs.js]: I would have gone for

```
export const NOTIFICATION_STATES = {
    ALL_MESSAGES_LOUD: 'all_messages_loud',
    ...
}
```

but this is fine too

File: src/RoomNotifs.js [hunk 13]
```diff
@@ -32,71 +39,84 @@ export function getRoomNotifsState(roomId) {
...
 
     const actionsObject = PushProcessor.actionListToActionsObject(roomRule.actions);
-    if (actionsObject.tweaks.sound) return 'all_messages_loud';
+    if (actionsObject.tweaks.sound) return ALL_MESSAGES_LOUD;
 
     return null;
 }
 
 export function setRoomNotifsState(roomId, newState) {
+    if (newState == 'mute') {
```
Comment (richvdh) [inline src/RoomNotifs.js]: s/'mute'/MUTE/

File: src/components/views/rooms/RoomTile.js [hunk 14]
```diff
@@ -179,15 +178,19 @@ module.exports = React.createClass({
...
+        const mentionBadges = this.props.highlight && this._shouldShowMentionBadge();
+        const badges = notifBadges || mentionBadges;
+
         var classes = classNames({
             'mx_RoomTile': true,
             'mx_RoomTile_selected': this.props.selected,
             'mx_RoomTile_unread': this.props.unread,
-            'mx_RoomTile_unreadNotify': notificationCount > 0 && !this.state.areNotifsMuted,
-            'mx_RoomTile_highlight': this.props.highlight,
+            'mx_RoomTile_unreadNotify': notificationCount > 0 && this._shouldShowNotifBadge(),
```
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: could just use `notifBadges` here

File: src/components/views/rooms/RoomTile.js [hunk 15]
```diff
@@ -179,15 +178,19 @@ module.exports = React.createClass({
...
+            'mx_RoomTile_highlight': this.props.highlight && this._shouldShowMentionBadge(),
```
Comment (richvdh) [inline src/components/views/rooms/RoomTile.js]: could just use `mentionBadges` here

File: src/RoomNotifs.js [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    const cli = MatrixClientPeg.get();
+    const promises = [];
+
+    // delete the room rule
+    const roomRule = cli.getRoomPushRule('global', roomId);
+    if (roomRule) {
+        promises.push(cli.deletePushRule('global', 'room', roomRule.rule_id));
+    }
+
+    // add an override rule to squelch everything in this room
```
Comment (richvdh) [inline src/RoomNotifs.js]: suggest rewording this slightly to make it clearer that this will replace existing rules. perhaps just "add/replace override rule ..."

**Comment by richvdh** [comment]
just some minor quibbling
